File: Flutter/lib/classes/Joints.dart
import 'dart:convert';

import 'package:flutter/material.dart';
import 'package:rob_bhit/utils/helper.dart';

class JointNotifier extends ValueNotifier<List<Joint>> {

  JointNotifier(super.value);

  void set(int i, Joint newValue) {

    super.value[i] = newValue;

  }

  void setList(List<Joint> newValue) {

    super.value = newValue;

  }

}

class Joint {

  const Joint(
    this.angle,
    this.turns,
    this.color,
    this.jointNo
  );
  
  final int jointNo;
  final double turns, angle;
  final Color color;

  String toJson() => {
    "angle": angle,
    "turns": turns,
    "color": color,
    "jointNo": jointNo
  }.toString();

  static Joint fromMap(int i, Map<String, dynamic> decodedJson) {

    return Joint(
      decodedJson["angle"] ?? 0.0,
      decodedJson["turns"] ?? 0.0,
      jointColors[i],
      i + 1
    );

  }

  static Joint fromJson(int i, String jsonStr) {

    Map<String, dynamic> decodedJson = (json.decode(jsonStr) as Map<String, dynamic>);

    return Joint.fromMap(i, decodedJson);

  }

}

File: Backend/data_extraction/Classes/joints.py
from json import loads
    
class Joint():

    def __init__(self, angle: float = 0.0, turns: float = 0.0):
        self.angle = angle
        self.turns = turns

    def set(self, angle: float, turns: float) -> "Joint":
        self.angle = angle
        self.turns = turns

        return self

    def toDict(self) -> dict:
        return {
            "angle": self.angle,
            "turns": self.turns
        }

    @staticmethod
    def fromJson(jsonStr: str) -> "Joint":
        return Joint.fromDict(loads(jsonStr))

    @staticmethod
    def fromDict(jointData: dict) -> "Joint":
        joints = Joint()
        joints.set(
            jointData["angle"],
            jointData["turns"]
        )

        return joints